codegen/module_system.go

```diff
@@ -0,0 +1,199 @@
+)
+
+// NewDefaultModuleSystem creates a fresh instance of the default zanzibar
+// module system (clients, endpoints)
+func NewDefaultModuleSystem(h *PackageHelper) (*module.System, error) {
+	system := module.NewSystem()
+	tmpl, err := NewTemplate(templateDir)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Register client module class and type generators
+	if err := system.RegisterClass("client", module.Class{
+		Directory:         "clients",
+		ClassType:         module.MultiModule,
+		ClassDependencies: []string{},
+	}); err != nil {
+		return nil, errors.Wrapf(err, "Error registering client class")
+	}
+
+	if err = system.RegisterClassType("client", "http", &HTTPClientGenerator{
+		templates:     tmpl,
+		packageHelper: h,
+	}); err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"Error registering http client class type",
+		)
+	}
+
+	// TODO: Register endpoint module class and type generators
+	return system, nil
+}
+
+/*
+ * HTTP Client Generator
+ */
+
+// HTTPClientGenerator generates an instance of a zanzibar http client
+type HTTPClientGenerator struct {
+	templates     *Template
+	packageHelper *PackageHelper
+}
+
+// Generate returns the HTTP client generated files as a map of relative file
+// path (relative to the target buid directory) to file bytes.
+func (generator *HTTPClientGenerator) Generate(
+	instance *module.Instance,
+) (map[string][]byte, error) {
+	// Parse the client config from the endpoint JSON file
+	clientConfig, err := generator.readConfig(instance.JSONFileRaw)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"Error reading HTTP client %s JSON config",
+			instance.InstanceName,
+		)
+	}
+
+	clientSpec, err := NewHTTPClientSpec(
+		filepath.Join(
+			instance.BaseDirectory,
+			instance.Directory,
+			instance.JSONFileName,
+		),
+		clientConfig,
+		generator.packageHelper,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"Error initializing HTTPClientSpec for %s",
+			instance.InstanceName,
+		)
+	}
+
+	clientMeta := &ClientMeta{
+		PackageName:      clientSpec.ModuleSpec.PackageName,
+		Services:         clientSpec.ModuleSpec.Services,
+		IncludedPackages: clientSpec.ModuleSpec.IncludedPackages,
+		ClientID:         clientSpec.ClientID,
+	}
+
+	// TODO: add helper to template to return byte array
+	clientBuffer := bytes.NewBuffer(nil)
+	clientWriter := bufio.NewWriter(clientBuffer)
+	if err := generator.templates.template.ExecuteTemplate(
+		clientWriter,
+		"http_client.tmpl",
+		clientMeta,
+	); err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"Error executing HTTP Client template for %s",
+			instance.InstanceName,
+		)
+	}
+	if err := clientWriter.Flush(); err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"Error generating HTTP client from template for %s",
+			instance.InstanceName,
+		)
+	}
+
+	structsBuffer := bytes.NewBuffer(nil)
+	structsWriter := bufio.NewWriter(structsBuffer)
+	if err := generator.templates.template.ExecuteTemplate(
```

Comment:
nitpick: We should have a consistent err pattern in the codebase.

Currently we both use

```
err = x();
if err != nil {
    return nil, err
}
```

and

```
if err := x(); err != nil {
    return nil, err
}
```